ports/nrf/mphalport.c

```diff
@@ -35,6 +35,82 @@
 #include "nrfx_errors.h"
 #include "nrfx_config.h"
 
+#if MICROPY_PY_TIME_TICKS
+#include "nrfx_rtc.h"
+#endif
+
+
+#if MICROPY_PY_TIME_TICKS
+// use RTC1 for time ticks generation
+// with 32kHz tick resolution and overflow
+// handling in irq
+
+nrfx_rtc_t rtc1 = NRFX_RTC_INSTANCE(1);
+
+volatile mp_uint_t rtc_overflows = 0;
+
+const nrfx_rtc_config_t rtc_config_time_ticks = {
+    .prescaler    = 0,
+    .reliable     = 0,
+    .tick_latency = 0,
+    #ifdef NRF51
+    .interrupt_priority = 1,
+    #else
+    .interrupt_priority = 3,
+    #endif
+};
+
+STATIC void rtc_irq_time(nrfx_rtc_int_type_t event) {
+    // irq handler for overflow
+    if (event == NRFX_RTC_INT_OVERFLOW) {
+        rtc_overflows += 1;
+    }
+}
+
+void rtc1_init_time_ticks(void) {
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
+    nrfx_rtc_overflow_enable(&rtc1, true);
+    nrfx_rtc_enable(&rtc1);
+}
+
+mp_uint_t mp_hal_ticks_ms(void) {
+    // note that COUNTER * 1000 / 32768 would overflow during calculation, so use
+    // the less obvious * 125 / 4096 calculation (overflow secure)
+    //
+    // also make sure not to call this function within an irq with higher
+    // prio than the RTC's irq. This would introduce the danger of 
+    // preempting the RTC irq and calling mp_hal_ticks_ms() at that time
+    // would return a false result
+    //
+    // also lets guard the function against interrupting from the RTC 
+    // overflow irq while the calculation takes place
+    mp_uint_t ticks;
+    rtc1.p_reg->INTENCLR = RTC_INTENSET_OVRFLW_Msk;
+    ticks = (rtc_overflows << 9) * 1000 + ((mp_uint_t)rtc1.p_reg->COUNTER * 125 / 4096);
```

Comment:
> Hmm.. why is the check for counter necessary? If overflow is set it can be safely assumed that overflows must be increased by 1?

If OVRFLW is set it could be before or after COUNTER is read.  If before then the adjustment must be made, if after then no adjustment.  The before case is when COUNTER is very small (because it just overflowed and was set to zero), the after case is when COUNTER is very large (because it's just about to overflow but we read it right before it overflows).  The extra check for counter is to distinguish these cases and I chose `1<<23` because it's halfway between min and max values of COUNTER.

> Also before enabling the irq we should reset the irq flag since the rtc_irq is otherwise immediately called & this increases the overflow counter again...



No.  The global `rtc_overflows` still needs to be incremented.  Only the local copy here is adjusted by 1 if needed.

> Also to simplify things.. Perhaps it just make sense to combine the code altogether and calculate a 64bit wide usec tick counter for both ticks_ms and ticks_us as you pointed out below. For `mp_hal_ticks_ms()` we'd just in addition divide by 1000 & mask by 0xffff ffff.

But a div-by-1000 of a uint64_t will be expensive.  I think it's ok to duplicate code like this, it's time critical and needs to be precise.

> This complexity makes me think if we shouldn't just reconsider and move to the timer's 32bit 1MHz clock?

If we want both ms and us then there needs to be some complexity somewhere, unless there are two 32-bit dedicated timers, one running at 1kHz for ms and one at 1MHz for us.

Reply:
thanks for the clarification, now I understand. Agree with argumentation. Will look into this (it's kind of getting really interesting here ;))

Reply:
done. I copy&pasted your comment above to the code if that's ok for you. I believe it clarifies the issue nicely.